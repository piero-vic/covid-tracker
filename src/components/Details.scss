.details-header {
  background-color: #4269b3;
  display: flex;
  align-items: center;
  padding: 0.6rem 0.4rem;

  p {
    margin: 0 auto;
  }
}

.main-card {
  p {
    font-size: 12px;
  }
}

.data-item {
  background-color: #4269b3;
  font-size: 12px;
  padding: 0.4rem;
  display: flex;
  justify-content: space-between;

  .data {
    font-weight: bold;
  }
}

img {
  max-width: 100%;
}
