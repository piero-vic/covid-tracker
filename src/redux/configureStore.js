import { createStore, combineReducers, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import dataReducer from './data/data';
import covidReducer from './covid/covid';

const reducer = combineReducers({
  data: dataReducer,
  covid: covidReducer,
});

const store = createStore(reducer, applyMiddleware(thunk, logger));

export default store;
