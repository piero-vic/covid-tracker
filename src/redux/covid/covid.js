const GET_COVID_DATA = 'covid/GET_COVID_DATA';
const CLEAR_DATA = 'covid/CLEAR_DATA';

const initialState = {};

const covidReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_COVID_DATA:
      return { ...state, [action.payload.id]: action.payload };
    case CLEAR_DATA:
      return {};
    default:
      return state;
  }
};

export const getCovidData = (payload) => ({
  type: GET_COVID_DATA,
  payload,
});

export const clearCovidData = () => ({
  type: CLEAR_DATA,
});

export const fetchCovidData = (slug) => async (dispatch) => {
  const date = new Date();
  const dateString = date.toISOString().slice(0, 10);
  const response = await fetch(
    `https://api.covid19tracking.narrativa.com/api/${dateString}/country/us/region/${slug.replace(
      '-',
      '_',
    )}`,
  );
  const allData = await response.json();
  const covidData = allData.dates[dateString].countries.US.regions[0];
  dispatch(
    getCovidData({
      id: covidData.id,
      confirmed: covidData.today_confirmed,
      new_confirmed: covidData.today_new_confirmed,
      deaths: covidData.today_deaths,
      new_deaths: covidData.today_new_deaths,
      open_cases: covidData.today_open_cases,
      new_open_cases: covidData.today_new_open_cases,
      recovered: covidData.today_recovered,
      new_recovered: covidData.today_new_recovered,
      tests: covidData.today_tests,
      new_tests: covidData.today_new_tests,
      total_hospitalised_patients: covidData.today_total_hospitalised_patients,
      new_total_hospitalised_patients:
        covidData.today_new_total_hospitalised_patients,
    }),
  );
};

export default covidReducer;
