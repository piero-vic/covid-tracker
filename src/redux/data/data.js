const GET_ALL_DATA = 'data/GET_ALL_DATA';

const initialState = [];

const dataReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_ALL_DATA:
      return action.payload;
    default:
      return state;
  }
};

export const getAllData = (payload) => ({
  type: GET_ALL_DATA,
  payload,
});

export const fetchAllData = () => async (dispatch) => {
  const response = await fetch(
    'https://civilserviceusa.github.io/us-states/data/states.json',
  );
  const states = await response.json();
  dispatch(getAllData(states));
};

export default dataReducer;
