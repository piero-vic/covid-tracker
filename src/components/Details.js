import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { MdArrowBackIosNew } from 'react-icons/md';
import './Details.scss';
import { fetchCovidData } from '../redux/covid/covid';

const Details = (props) => {
  const { region } = props;

  const slug = region.slug.replace('-', '_');

  const covidData = useSelector((state) => state.covid);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!covidData[slug]) dispatch(fetchCovidData(slug));
  }, []);

  return (
    <div>
      <div className="details-header">
        <Link to="/">
          <MdArrowBackIosNew />
        </Link>
        <p>state/region details</p>
      </div>
      <div className="main-card">
        <img src={region.map_image_url} alt={`${region.state}-map`} />
        <div>
          <h1>{`${region.state} (${region.code})`}</h1>
          <p>{region.nickname}</p>
        </div>
      </div>
      <div className="bar">GENERAL INFORMATION</div>
      <ul>
        <li className="data-item">
          <span className="tag">Population</span>
          <span className="data">
            {region.population.toLocaleString('en-US')}
          </span>
        </li>
        <li className="data-item">
          <span className="tag">Population Rank</span>
          <span className="data">{region.population_rank}</span>
        </li>
        <li className="data-item">
          <span className="tag">Capital City</span>
          <span className="data">{region.capital_city}</span>
        </li>
      </ul>
      <div className="bar">TODAY&apos;S COVID INFORMATION</div>
      <ul>
        {covidData[slug] ? (
          Object.keys(covidData[slug])
            .filter((x) => x !== 'id')
            .map((key) => (
              <li key={key} className="data-item">
                <span className="tag">
                  {`${key.charAt(0).toUpperCase()}${key
                    .slice(1)
                    .replace(/_/g, ' ')}`}
                </span>
                <span className="data">
                  {covidData[slug][key].toLocaleString('en-US')}
                </span>
              </li>
            ))
        ) : (
          <li className="data-item">Loading...</li>
        )}
      </ul>
    </div>
  );
};

Details.propTypes = {
  region: PropTypes.instanceOf(Object).isRequired,
};

export default Details;
