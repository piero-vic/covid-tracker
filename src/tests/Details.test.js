import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Details from '../components/Details';

import data from './__mocks__/data_mock.json';
import covid from './__mocks__/covid_mock.json';

describe('Details test', () => {
  const initialState = { data, covid };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  test('Renders correctly', () => {
    const detailsComponent = render(
      <Provider store={store}>
        <Router>
          <Details region={data[0]} />
        </Router>
      </Provider>,
    );
    expect(detailsComponent).toMatchSnapshot();
  });

  test('Renders two lists of information', () => {
    const detailsComponent = render(
      <Provider store={store}>
        <Router>
          <Details region={data[0]} />
        </Router>
      </Provider>,
    );
    const [generalInformation, covidInformation] = detailsComponent.container.querySelectorAll('ul');
    expect(generalInformation.children).toHaveLength(3);
    expect(covidInformation.children).toHaveLength(12);
  });

  test('Renders information correctly', () => {
    const detailsComponent = render(
      <Provider store={store}>
        <Router>
          <Details region={data[0]} />
        </Router>
      </Provider>,
    );

    const populationNumber = detailsComponent.getByText('Population').nextSibling;
    expect(populationNumber.textContent).toBe('4,833,722');

    const confirmed = detailsComponent.getByText('Confirmed').nextSibling;
    expect(confirmed.textContent).toBe('863,170');
  });
});
