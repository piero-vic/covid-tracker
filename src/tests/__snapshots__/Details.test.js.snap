// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details test Renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="details-header"
        >
          <a
            href="/"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17.77 3.77L16 2 6 12l10 10 1.77-1.77L9.54 12z"
              />
            </svg>
          </a>
          <p>
            state/region details
          </p>
        </div>
        <div
          class="main-card"
        >
          <img
            alt="Alabama-map"
            src="https://cdn.civil.services/us-states/maps/alabama-large.png"
          />
          <div>
            <h1>
              Alabama (AL)
            </h1>
            <p>
              Yellowhammer State
            </p>
          </div>
        </div>
        <div
          class="bar"
        >
          GENERAL INFORMATION
        </div>
        <ul>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Population
            </span>
            <span
              class="data"
            >
              4,833,722
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Population Rank
            </span>
            <span
              class="data"
            >
              23
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Capital City
            </span>
            <span
              class="data"
            >
              Montgomery
            </span>
          </li>
        </ul>
        <div
          class="bar"
        >
          TODAY'S COVID INFORMATION
        </div>
        <ul>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Confirmed
            </span>
            <span
              class="data"
            >
              863,170
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Deaths
            </span>
            <span
              class="data"
            >
              16,402
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New confirmed
            </span>
            <span
              class="data"
            >
              2,060
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New deaths
            </span>
            <span
              class="data"
            >
              22
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New open cases
            </span>
            <span
              class="data"
            >
              2,038
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New recovered
            </span>
            <span
              class="data"
            >
              0
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New tests
            </span>
            <span
              class="data"
            >
              0
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              New total hospitalised patients
            </span>
            <span
              class="data"
            >
              0
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Open cases
            </span>
            <span
              class="data"
            >
              551,078
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Recovered
            </span>
            <span
              class="data"
            >
              295,690
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Tests
            </span>
            <span
              class="data"
            >
              0
            </span>
          </li>
          <li
            class="data-item"
          >
            <span
              class="tag"
            >
              Total hospitalised patients
            </span>
            <span
              class="data"
            >
              0
            </span>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="details-header"
      >
        <a
          href="/"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M17.77 3.77L16 2 6 12l10 10 1.77-1.77L9.54 12z"
            />
          </svg>
        </a>
        <p>
          state/region details
        </p>
      </div>
      <div
        class="main-card"
      >
        <img
          alt="Alabama-map"
          src="https://cdn.civil.services/us-states/maps/alabama-large.png"
        />
        <div>
          <h1>
            Alabama (AL)
          </h1>
          <p>
            Yellowhammer State
          </p>
        </div>
      </div>
      <div
        class="bar"
      >
        GENERAL INFORMATION
      </div>
      <ul>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Population
          </span>
          <span
            class="data"
          >
            4,833,722
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Population Rank
          </span>
          <span
            class="data"
          >
            23
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Capital City
          </span>
          <span
            class="data"
          >
            Montgomery
          </span>
        </li>
      </ul>
      <div
        class="bar"
      >
        TODAY'S COVID INFORMATION
      </div>
      <ul>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Confirmed
          </span>
          <span
            class="data"
          >
            863,170
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Deaths
          </span>
          <span
            class="data"
          >
            16,402
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New confirmed
          </span>
          <span
            class="data"
          >
            2,060
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New deaths
          </span>
          <span
            class="data"
          >
            22
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New open cases
          </span>
          <span
            class="data"
          >
            2,038
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New recovered
          </span>
          <span
            class="data"
          >
            0
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New tests
          </span>
          <span
            class="data"
          >
            0
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            New total hospitalised patients
          </span>
          <span
            class="data"
          >
            0
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Open cases
          </span>
          <span
            class="data"
          >
            551,078
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Recovered
          </span>
          <span
            class="data"
          >
            295,690
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Tests
          </span>
          <span
            class="data"
          >
            0
          </span>
        </li>
        <li
          class="data-item"
        >
          <span
            class="tag"
          >
            Total hospitalised patients
          </span>
          <span
            class="data"
          >
            0
          </span>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
